nservicebus/nhibernate/saga-concurrency.md

```diff
@@ -15,7 +16,7 @@ One of the most critical things about persistence of sagas is proper concurrency
 
 As stated in [saga concurrency](/nservicebus/sagas/concurrency.md), the saga persistence system depends on the data access providing an optimistic approach to concurrency. With NHibernate this results in appending a `WHERE` clause containing all known values of saga data fields when doing `UPDATE`s. This ensures that the saga data is still in the same state as when it was read.
 
-This approach has a downside of very poor performance in high-contention scenarios where a single saga is accessed by multiple message-processing threads. In order to overcome this starting from Version 4.1.0 the NHibernate saga persister uses **additional** pessimistic concurrency control using `UPDLOCK` hint. A lock is created when fetching the saga instance from the database and the lock is held till the end of the transaction preventing other threads from fetching that particular saga. The advantage of this approach that it minimizes the number of retries that would be caused should only the optimistic concurrency was employed. Multiple saga instances can still be processed in parallel.
+This approach has a downside of very poor performance in high-contention scenarios where a single saga is accessed by multiple message-processing threads. These threads read the same saga state and process their messages but only one can even succeed persisting the new state. Others experience concurrency violation error and have to try again. In order to overcome this the NHibernate saga persister uses **additional** pessimistic concurrency control using `UPDLOCK` hint. A lock is created when fetching the saga instance from the database and is held till the end of the transaction blocking other threads that try fetching that particular saga. The advantage of this approach that it minimizes the number of retries that would be caused should only the optimistic concurrency was employed. Multiple saga instances can still be processed in parallel.
```

Comment:
>This approach has a downside of very poor performance in high-contention scenarios where a single saga is accessed by multiple message-processing threads.

This approach has a downside of ~~very~~ poor performance in high-contention scenarios where a single saga is accessed by multiple message-processing threads.


> These threads read the same saga state and process their messages but only one can even succeed persisting the new state. 

These threads read the same saga state and process their messages but only one can ~~even~~ succeed *in* persisting the new state. 

>Others experience concurrency violation error and have to try again.

Other threads processing the same message will experience concurrency violation error and will be retried.

>In order to overcome this the NHibernate saga persister uses **additional** pessimistic concurrency control using `UPDLOCK` hint.

this -> what does this imply? the concurrency exceptions? 

>The advantage of this approach that it minimizes the number of retries that would be caused should only the optimistic concurrency was employed. Multiple saga instances can still be processed in parallel.

Not clear. Can you clarify?